controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +141,129 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+	// In this case the method should still be allowed to continue.
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err := r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
+	m, err := r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+	}
+	clusterState.machineDeployments = m
+	clusterState.cluster = cluster
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterTopologyReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrap(err, "InfrastructureCluster")
+	}
+	return infra, nil
+}
+
+// getCurrentControlPlaneState returns information on the ControlPlane being used by the Cluster; If a references is not found,
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterTopologyReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "ControlPlane")
+	}
+
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the clusterClass of the given cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "ControlPlane")
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "ControlPlane ")
```

Comment:
```suggestion
		return nil, errors.Wrap(err, "ControlPlane ")
```
nit (same as we had with the MD below)